controllers/dnspolicy_status_updater.go

```diff
@@ -0,0 +1,124 @@
+	return &DNSPolicyStatusUpdater{client: client}
+}
+
+type DNSPolicyStatusUpdater struct {
+	client *dynamic.DynamicClient
+}
+
+func (r *DNSPolicyStatusUpdater) Subscription() controller.Subscription {
+	return controller.Subscription{
+		ReconcileFunc: r.updateStatus,
+		Events: []controller.ResourceEventMatcher{
+			{Kind: &machinery.GatewayGroupKind},
+			{Kind: &kuadrantv1alpha1.DNSPolicyGroupKind},
+			{Kind: &DNSRecordGroupKind},
+		},
+	}
+}
+
+func (r *DNSPolicyStatusUpdater) updateStatus(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
+	logger := controller.LoggerFromContext(ctx).WithName("DNSPolicyStatusUpdater")
+
+	policyTypeFilterFunc := dnsPolicyTypeFilterFunc()
+	policyAcceptedFunc := dnsPolicyAcceptedStatusFunc(state)
+
+	policies := lo.FilterMap(topology.Policies().Items(), policyTypeFilterFunc)
+
+	logger.V(1).Info("updating dns policy statuses", "policies", len(policies))
+
+	for _, policy := range policies {
+		pLogger := logger.WithValues("policy", policy.Name)
+
+		if policy.GetDeletionTimestamp() != nil {
+			pLogger.V(1).Info("policy marked for deletion, skipping", "name", policy.Name, "namespace", policy.Namespace)
+			continue
+		}
+
+		// copy initial conditions, otherwise status will always be updated
+		newStatus := &kuadrantv1alpha1.DNSPolicyStatus{
+			Conditions:         slices.Clone(policy.Status.Conditions),
+			ObservedGeneration: policy.Status.ObservedGeneration,
+		}
+
+		accepted, err := policyAcceptedFunc(policy)
+		meta.SetStatusCondition(&newStatus.Conditions, *kuadrant.AcceptedCondition(policy, err))
+
+		// do not set enforced condition if Accepted condition is false
+		if !accepted {
+			meta.RemoveStatusCondition(&newStatus.Conditions, string(kuadrant.PolicyConditionEnforced))
+		} else {
+			policyRecords := lo.FilterMap(topology.Objects().Items(), func(item machinery.Object, _ int) (*kuadrantdnsv1alpha1.DNSRecord, bool) {
+				if rObj, isObj := item.(*controller.RuntimeObject); isObj {
+					if record, isRec := rObj.Object.(*kuadrantdnsv1alpha1.DNSRecord); isRec {
+						return record, lo.ContainsBy(topology.Policies().Parents(item), func(item machinery.Policy) bool {
+							return item.GetLocator() == policy.GetLocator()
+						})
+					}
+				}
+				return nil, false
+			})
+
+			enforcedCond := enforcedCondition(policyRecords, policy)
+			meta.SetStatusCondition(&newStatus.Conditions, *enforcedCond)
+
+			//ToDo: Deal with messages, these should probably be retrieved from state after the reconciliation task
+			// add some additional user friendly context
+			//if errors.Is(specErr, ErrNoAddresses) && !strings.Contains(eCond.Message, ErrNoAddresses.Error()) {
+			//	eCond.Message = fmt.Sprintf("%s : %s", eCond.Message, ErrNoAddresses.Error())
+			//}
+			//if errors.Is(specErr, ErrNoRoutes) && !strings.Contains(eCond.Message, ErrNoRoutes.Error()) {
+			//	eCond.Message = fmt.Sprintf("%s : %s", eCond.Message, ErrNoRoutes)
+			//}
```

Comment:
@maleck13 These more specific error messages are not currently being put into the status of the policy. I have an idea of how to do it, but i don't think this needs to block this PR, i can create a follow on task to fix this after merge.

Reply:
This is updated now